firmware/src/services/iot/cloud/cloud_service.c

```diff
@@ -0,0 +1,628 @@
+        {
+            int optVal = 1;
+
+            ret = BSD_setsockopt(*context->tcpClientSocket,
+                                 SOL_SSL_SOCKET,
+                                 SO_SSL_ENABLE_SNI_VALIDATION,
+                                 &optVal,
+                                 sizeof(optVal));
+        }
+
+        if (ret == BSD_SUCCESS)
+        {
+            ret = BSD_connect(*context->tcpClientSocket,
+                              (struct bsd_sockaddr*)&addr,
+                              sizeof(struct bsd_sockaddr_in));
+        }
+        else
+        {
+            debug_printError("CLOUD: Socket connect failed");
+            LED_SetRed(LED_STATE_BLINK_SLOW);
+            shared_networking_params.haveERROR = 1;
+            BSD_close(*context->tcpClientSocket);
+        }
+    }
+    return ret;
+}
+
+//
+// Initiates MQTT connection
+//
+static void connectMQTT()
+{
+    time_t    currentTime;   // = time(NULL);
+    struct tm sys_time;
+
+    RTC_RTCCTimeGet(&sys_time);
+    currentTime = mktime(&sys_time);
+
+    debug_printTrace("CLOUD: Sending MQTT CONNECT at %s", ctime(&currentTime));
+
+    if (currentTime > 0)
+    {
+        pf_mqtt_client->MQTT_CLIENT_connect(ateccsn);
+    }
+
+    waitingForMQTT = true;
+    debug_printInfo("CLOUD: Starting MQTT Timeout");
+    mqttTimeoutTaskHandle = SYS_TIME_CallbackRegisterMS(mqttTimeoutTaskcb, 0, CLOUD_MQTT_TIMEOUT_COUNT, SYS_TIME_SINGLE);
+
+    // MQTT SUBSCRIBE packet will be sent after the MQTT connection is established.
+    sendSubscribe = true;
+}
+
+//
+// Calls MQTT SUBSCRIBE
+//
+void CLOUD_subscribe(void)
+{
+    if (shared_networking_params.haveMqttConnection != 1)
+    {
+        debug_printError("CLOUD: MQTT not connected");
+        return;
+    }
+
+    debug_printTrace("CLOUD: Sending MQTT SUBSCRIBE");
+    if (pf_mqtt_client->MQTT_CLIENT_subscribe() == true)
+    {
+        sendSubscribe = false;
+    }
+}
+
+//
+// Initiates MQTT DISCONNECT
+// This forces a reconnect
+//
+void CLOUD_disconnect(void)
+{
+    if (MQTT_GetConnectionState() == CONNECTED)
+    {
+        debug_printWarn("CLOUD: Sending MQTT DISCONNECT");
+        MQTT_Disconnect(MQTT_GetClientConnectionInfo());
+    }
+}
+
+// Todo: This declaration supports the hack below
+packetReceptionHandler_t* getSocketInfo(uint8_t sock);
+
+
+void CLOUD_task(void)
+{
+    mqttContext*  mqttConnnectionInfo = MQTT_GetClientConnectionInfo();
+    socketState_t socketState         = BSD_GetSocketState(*mqttConnnectionInfo->tcpClientSocket);
+
+    // typedef enum
+    // {
+    // 	NOT_A_SOCKET = 0,           // This is not a socket
+    // 	SOCKET_CLOSED,              // Socket closed
+    // 	SOCKET_IN_PROGRESS,         // The TCP listen or initiate a connection
+    // 	SOCKET_CONNECTED,           // The TCP is in established state user can send/receive data
+    // 	SOCKET_CLOSING              // The user initiate the closing procedure for this socket
+    // } socketState_t;

```

Comment:
needed?

Reply:
This was more for note to help debugging the socketState state machine.  Removed.